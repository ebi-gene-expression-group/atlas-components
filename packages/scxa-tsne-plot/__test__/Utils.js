import HighchartsSeriesGenerator from 'highcharts-series-generator'

const seriesNames = [`0`, `1`, `2`, `3`, `4`]
const maxPointsPerSeries = 1000

const gradientColourRanges = () => {
  return [
    {
      colour: `rgb(215, 255, 255)`,
      threshold: 0,
      stopPosition: 0
    },
    {
      colour: `rgb(128, 255, 255)`,
      threshold: 10,
      stopPosition: 20
    },
    {
      colour: `rgb(0, 85, 225)`,
      threshold: 100,
      stopPosition: 40
    },
    {
      colour: `rgb(0, 0, 115)`,
      threshold: 10000,
      stopPosition: 100
    }
  ]
}

const plotData = (series) => {
  const allPoints = series.reduce((acc, series) => acc.concat(series.data), [])

  return {
    series: series,
    min: Math.min(...allPoints.map((point) => point.expressionLevel)),
    max: Math.max(...allPoints.map((point) => point.expressionLevel)),
    unit: `TPM`
  }
}

const randomHighchartsSeries = () => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithNamesAndMaxPoints = (seriesNames, maxPointsPerSeries) => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithSeed = () => {
  const seed = `A hair, Morty. I need one of your hairs. This isnâ€™t Game of Thrones.`

  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries, seed)
}

export { plotData, gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithSeed, randomHighchartsSeriesWithNamesAndMaxPoints}