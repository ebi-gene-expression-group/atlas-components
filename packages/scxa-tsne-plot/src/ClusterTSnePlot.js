import React from 'react'
import PropTypes from 'prop-types'
import Color from 'color'

import ScatterPlotLoader from './plotloader/PlotLoader'

const _colourizeClusters = (highlightSeries) =>
  (series) => series.map((aSeries) => {
    if (!highlightSeries.length || highlightSeries.includes(aSeries.name)) {
      return aSeries
    } else {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => ({
          ...point,
          color: Color(`lightgrey`).alpha(0.65).rgb().toString()
        }))
      }
    }
  })

const ClusterTSnePlot = (props) => {
  const {availableClusters, k, perplexityArray, perplexity} = props                     // Select
  const {plotData, highlightClusters, height, onChange, onChangePerplexity} = props     // Chart
  const {loading, resourcesUrl, errorMessage} = props                                   // Overlay

  const highchartsConfig = {
    plotOptions: {
      scatter: {
        tooltip: {
          headerFormat: `<b>Cluster {series.name}</b><br>`,
          pointFormat: `{point.name}`
        }
      }
    },
    colors: [
              'rgba(178, 95, 188, 0.7)',
              'rgba(118, 179, 65, 0.7)',
              'rgba(104, 130, 207, 0.7)',
              'rgba(206, 155, 68, 0.7)',
              'rgba(200, 87, 123, 0.7)',
              'rgba(79, 174, 132, 0.7)',
              'rgba(201, 92, 63, 0.7)',
              'rgba(124, 127, 57, 0.7)'
            ],
    chart: {
      height: height
    }
  }

  const clusterOptions = availableClusters.sort().map((name, ix) => (
    <option key={ix} value={name}>{name}</option>
  ))

  const perplexityOptions = perplexityArray.sort().map((name, ix) => (
    <option key={ix} value={name}>{name}</option>
  ))

  return [
      <div className={`row`}>
          <div key={`cluster-select`} className={`column medium-6`}>
            <label>Number of clusters, <i>k</i></label>
            <select value={k} onChange={onChange}>
              {clusterOptions}
            </select>
          </div>

          <div key={`perplexity-select`} className={`column medium-6`}>
              <label>Perplexity, <i>perplexity</i></label>
              <select value={perplexity} onChange={onChangePerplexity}>
                  {perplexityOptions}
              </select>
          </div>
      </div>,

      <ScatterPlotLoader key={`cluster-plot`}
                         wrapperClassName={`row column`}
                         series={_colourizeClusters(highlightClusters)(plotData.series)}
                         highchartsConfig={highchartsConfig}
                         loading={loading}
                         resourcesUrl={resourcesUrl}
                         errorMessage={errorMessage}
      />
  ]
}

ClusterTSnePlot.propTypes = {
  height: PropTypes.number.isRequired,

  plotData: PropTypes.shape({
    series: PropTypes.array.isRequired
  }),
  highlightClusters: PropTypes.array,

  availableClusters: PropTypes.array.isRequired,
  k: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,

  loading: PropTypes.bool.isRequired,
  resourcesUrl: PropTypes.string,
  errorMessage: PropTypes.string
}

export {ClusterTSnePlot as default, _colourizeClusters}
