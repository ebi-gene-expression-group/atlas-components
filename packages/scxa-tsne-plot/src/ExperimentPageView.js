import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'
import queryString from 'query-string'

import {_colourizeClusters, _colourizeExpressionLevel} from './SeriesMapper'
import TSnePlot from 'expression-atlas-tsne-plot'
import AtlasAutocomplete from 'atlas-autocomplete'

class ExperimentPageView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      params: queryString.parse(props.location.search),
      geneId: ``,
      k: 0,
      clusterId: [],
    }
  }

  handleChange(param, item) {
    let _newparam = {}
    _newparam[param] = (param === `k` ? Number.parseInt(item.target.value) : item)
    this.setState(_newparam)

    this.props.history.push(
      `?` +
      queryString.stringify({
        geneId: (param === `geneId` ? item : this.state.geneId),
        k: (param === `k` ? item.target.value : this.state.k)})
      + `&clusterId=[${this.state.clusterId}]`
    )
  }

  componentDidMount() {
    this.setState({
      geneId: this.state.params.geneId,
      k: this.state.params.k !== undefined ? Number.parseInt(this.state.params.k) : this.state.k,
      clusterId: JSON.parse(this.state.params.clusterId || `[]`)
    })
  }

  render() {
    const clusterOptions = this.props.availableClusters.sort()
      .map((name, ix) => (
        <option key={ix} value={name}>{name}</option>
      ))

    const selectedCluster = this.state.k ? this.state.k : this.props.availableClusters[0]

    const clusterHighchartsConfig = {
      plotOptions: {
        scatter: {
          tooltip: {
            headerFormat: `<b>Cluster {series.name}</b><br>`,
            pointFormat: `{point.name}`
          }
        }
      }
    }

    const geneExpressionHighchartsConfig = {
      plotOptions: {
        scatter: {
          tooltip: {
            headerFormat: `<b>{point.key} â€“ Cluster {series.name}</b><br>`,
            pointFormat: `Expression level: {point.expressionLevel}`
          }
        }
      }
    }

    return (
      <div className={`row`}>

        <div className={`small-6 columns`}>
          <label>Clustering: {this.state.k}</label>
          <select value={selectedCluster} onChange={this.handleChange.bind(this, `k`)}>
            {clusterOptions}
          </select>
          <TSnePlot atlasUrl={this.props.atlasUrl}
                    sourceUrl={`json/experiments/${this.props.experimentAccession}/tsneplot/clusters/${this.state.k}`}
                    seriesMapper={_colourizeClusters(this.props.highlightClusters)}
                    resourcesUrl={this.props.resourcesUrl}
                    highchartsConfig={clusterHighchartsConfig}/>
        </div>

        <div className={`small-6 columns`}>
          <AtlasAutocomplete atlasUrl={this.props.atlasUrl}
                             suggesterEndpoint={this.props.suggesterEndpoint}
                             enableSpeciesFilter={false}
                             initialValue={this.state.geneId}
                             onSelect={this.handleChange.bind(this, `geneId`)}/>
          <TSnePlot atlasUrl={this.props.atlasUrl}
                    sourceUrl={`json/experiments/${this.props.experimentAccession}/tsneplot/expression/${this.state.geneId}`}
                    seriesMapper={_colourizeExpressionLevel(this.props.expressionColourHue)}
                    resourcesUrl={this.props.resourcesUrl}
                    highchartsConfig={geneExpressionHighchartsConfig}/>
        </div>

      </div>
    )
  }
}

ExperimentPageView.propTypes = {
  props: PropTypes.object,
  atlasUrl: PropTypes.string,
  resourcesUrl: PropTypes.string,
  suggesterEndpoint: PropTypes.string,
  availableClusters: PropTypes.array.isRequired,
  highlightClusters: PropTypes.array,
  expressionColourHue: PropTypes.number
}

export default ExperimentPageView
