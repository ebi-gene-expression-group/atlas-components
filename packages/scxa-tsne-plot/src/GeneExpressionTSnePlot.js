import React from 'react'
import PropTypes from 'prop-types'
import Color from 'color'

import ScatterPlotLoader from './plotloader/PlotLoader'
import AtlasAutocomplete from 'expression-atlas-autocomplete'

import './util/MathRound'

const MAX_WHITE = 90


const _colourize = (colourStrings, defaultColour = `blue`, alpha = 0.65) => {
  const colours = colourStrings.map((colourStr) => Color(colourStr))

  return (val) => {
    if (isNaN(val)) {
      return Color(defaultColour).alpha(alpha).rgb().toString()
    }

    const bucket = val <= 0 ? 0 : val >= 1 ? colours.length - 2 : Math.floor(val / (1 / (colours.length - 1)))

    const loColour = colours[bucket]
    const hiColour = colours[bucket + 1]

    const redDelta = hiColour.red() - loColour.red()
    const greenDelta = hiColour.green() - loColour.green()
    const blueDelta = hiColour.blue() - loColour.blue()

    return Color(
      `rgb(` +
      `${Math.floor(loColour.red() + redDelta * val)}, ` +
      `${Math.floor(loColour.green() + greenDelta * val)}, ` +
      `${Math.floor(loColour.blue() + blueDelta * val)})`).alpha(alpha).rgb().toString()
  }
}


const _colourizeExpressionLevel = (gradientColours, highlightSeries) => {
  const colourize = _colourize(gradientColours)

  return (plotData) => plotData.series.map((aSeries) => {
    // I can’t think of a better way to reconcile series.name being a string and highlightSeries being an array of
    // numbers. For more flexibility we might think of having our series be identified by an arbitrary ID string
    if (!highlightSeries.length || highlightSeries.map((hs) => String(hs)).includes(aSeries.name)) {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => {
          return {
            ...point,
            expressionLevel: Math.round10(point.expressionLevel, -2),
            color: colourize(1 - (point.expressionLevel - plotData.min) / (plotData.max - plotData.min))
          }
        })
      }
    } else {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => ({
          ...point,
          expressionLevel: Math.round10(point.expressionLevel, -2),
          color: Color(`lightgrey`).alpha(0.65).rgb().toString()
        }))
      }
    }
  })
}

const GeneExpressionScatterPlot = (props) => {
  const {atlasUrl, suggesterEndpoint, geneId, onSelectGeneId} = props       // Suggester
  const {height, plotData, expressionGradientColours, highlightClusters} = props  // Chart
  const {loading, resourcesUrl, errorMessage} = props                       // Overlay

  const highchartsConfig = {
    plotOptions: {
      scatter: {
        tooltip: {
          headerFormat: `<b>{point.key} – Cluster {series.name}</b><br>`,
          pointFormat: geneId ? `Expression level: {point.expressionLevel} ${plotData.unit}` : `No gene selected`
        }
      }
    },
    chart: {
      height: height
    },
    title: {
      text: `Gene expression`
    }
  }

  const renderGradient = plotData.max && plotData.min && plotData.max > plotData.min
  const chartClassName = renderGradient ? `small-10 columns` : `small-12 columns`
  const gradient = renderGradient ?
    <LinearGradient height={height}
                    colours={expressionGradientColours}
                    plotData={plotData}/> :
    null

  return [
    <AtlasAutocomplete key={`expression-autocomplete`}
                       wrapperClassName={`row column`}
                       atlasUrl={atlasUrl}
                       suggesterEndpoint={suggesterEndpoint}
                       enableSpeciesFilter={false}
                       initialValue={geneId}
                       onSelect={ (event) => { onSelectGeneId(event) } }
    />,

    <ScatterPlotLoader key={`expression-plot`}
                       wrapperClassName={`row`}
                       chartClassName={chartClassName}
                       series={_colourizeExpressionLevel(expressionGradientColours, highlightClusters)(plotData)}
                       highchartsConfig={highchartsConfig}
                       children={gradient}
                       loading={loading}
                       resourcesUrl={resourcesUrl}
                       errorMessage={errorMessage}
    />
  ]
}

const LinearGradient = ({height, colours, plotData}) => {
  const background = colours.map((colour) => Color(colour).rgb().toString()).join(`, `)

  return (
    <div className={`small-2 columns text-center`}>
      <div><small>{Math.round10(plotData.max, -2)} {plotData.unit}</small></div>
      <div style={{width: `20px`, height: `${height - 100}px`, background: `linear-gradient(${background})`, verticalAlign: `middle`, margin: `auto`}}/>
      <div><small>{Math.round10(plotData.min, -2)} {plotData.unit}</small></div>
    </div>
  )
}

LinearGradient.propTypes = {
  height: PropTypes.number.isRequired,
  colours: PropTypes.arrayOf(PropTypes.string).isRequired,
  plotData: PropTypes.shape({
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    unit: PropTypes.string.isRequired
  })
}


GeneExpressionScatterPlot.propTypes = {
  height: PropTypes.number.isRequired,

  plotData: PropTypes.shape({
    series: PropTypes.array.isRequired,
    unit: PropTypes.string.isRequired,
    max: PropTypes.number,
    min: PropTypes.number
  }),
  expressionGradientColours: PropTypes.arrayOf(PropTypes.string).isRequired,
  highlightClusters: PropTypes.array,

  atlasUrl: PropTypes.string.isRequired,
  suggesterEndpoint: PropTypes.string.isRequired,
  onSelectGeneId: PropTypes.func.isRequired,

  loading: PropTypes.bool.isRequired,
  resourcesUrl: PropTypes.string,
  errorMessage: PropTypes.string
}

GeneExpressionScatterPlot.defaultProps = {
  expressionGradientColours: [`rgb(0, 0, 115)`, `rgb(0, 85, 225)`, `rgb(128, 255, 255)`, `rgb(215, 255, 255)`]
}

export {GeneExpressionScatterPlot as default, _colourizeExpressionLevel}
