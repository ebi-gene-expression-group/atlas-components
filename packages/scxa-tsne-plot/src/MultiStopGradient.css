.gradient {
    width: 20px;
    vertical-align: middle;
    margin: auto;
}

.tick {
    position: absolute;
    height: 2px;
    width: 20px;
}

.tickLabel {
    position: absolute;
    margin-top: -12px;
}

.left {
    composes: tickLabel;
    right: 20px;
    padding-right: 5px;
}

.right {
    composes: tickLabel;
    left: 20px;
    padding-left: 5px;
}