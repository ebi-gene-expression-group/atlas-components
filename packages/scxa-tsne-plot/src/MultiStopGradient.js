import React from 'react'
import PropTypes from 'prop-types'

import ScientificNotationNumber from 'expression-atlas-number-format'

import './util/MathRound'

// ranges have at least threshold (in ascending order) and stopPosition [0..100]
const putInRange = (ranges, val) => {
  if (val <= ranges[0].threshold) {
    return 0
  }

  if (val >= ranges[ranges.length - 1].threshold) {
    return 1
  }

  const rangeIndex = ranges.findIndex((range) => range.threshold >= val) - 1

  const withinRangeOffset =
    (val - ranges[rangeIndex].threshold) / (ranges[rangeIndex + 1].threshold - ranges[rangeIndex].threshold)

  return ranges[rangeIndex].stopPosition * (1 + withinRangeOffset) / 100
}



const MultiStopGradient = ({height, colourRanges, plotData}) => {
  const bg = colourRanges.map((colourRange) => `${colourRange.colour} ${colourRange.stopPosition}%`).join(`, `)
  const minMaxExpressionTickHeight = 2
  const gradientHeight = height - 100

  const minExpressionBottom =
    Math.min(
      gradientHeight - minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.min))
  const maxExpressionBottom =
    Math.max(
      minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.max))

  return (
    <div className={`small-2 columns text-center`}>
      <div>
        <small><ScientificNotationNumber value={colourRanges[colourRanges.length - 1].threshold}/> {plotData.unit}</small>
      </div>

      <div
        style={{
        width: `20px`,
        height: `${gradientHeight}px`,
        background: `linear-gradient(0deg, ${bg})`,
        verticalAlign: `middle`,
        margin: `auto`}}>

        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${maxExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.max, -2)}/></small>
          </div>
        </div>
        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${minExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.min, -2)}/></small>
          </div>
        </div>

      </div>

      <div>
        <small><ScientificNotationNumber value={colourRanges[0].threshold}/> {plotData.unit}</small>
      </div>
    </div>
  )
}

MultiStopGradient.propTypes = {
  height: PropTypes.number.isRequired,
  colourRanges: PropTypes.arrayOf(PropTypes.shape({
    colour: PropTypes.string.isRequired,
    threshold: PropTypes.number.isRequired
  })).isRequired,
  plotData: PropTypes.shape({
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    unit: PropTypes.string.isRequired
  })
}

export default MultiStopGradient
