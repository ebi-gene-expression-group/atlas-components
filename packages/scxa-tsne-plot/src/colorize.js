import Color from 'color'

// const defaultColor = `#eeeeee`
// const stops = [
//   // The first stop acts as a threshold: anything lower than this gets the defaultColor
//   {
//     cutoff: 0.1,
//     color: `#e1e9f4`
//   },
//   {
//     cutoff: 1,
//     color: `#d4e4fb`
//   },
//   {
//     cutoff: 10,
//     color: `#95adde`
//   },
//   {
//     cutoff: 100,
//     color: `#6077bf`
//   },
//   {
//     cutoff: 1000,
//     color: `#1151d1`
//   },
//   {
//     cutoff: 10000,
//     color: `#35419b`
//   },
//   {
//     cutoff: 100000,
//     color: `#0e0573`
//   },
//   {
//     cutoff: 1000000,
//     color: `#07004c`
//   }
// ]

const colorize = (stops, defaultColour) =>
  (val) => {
    // Below minimum
    if (val < stops[0].cutoff) {
      return Color(defaultColour).alpha(0.65).toString()
    }
    // Above maximum
    if (val > stops[stops.length - 1].cutoff) {
      return Color(stops[stops.length - 1].color).toString()
    }

    // General case
    const stopIndex = stops.findIndex(stop => val <= stop.cutoff)

    const minColor = Color(stops[Math.max(0, stopIndex - 1)].color)  // Because stopIndex may be 0
    const maxColor = Color(stops[stopIndex].color)

    // Transform colors to RGB
    const positionWithinStops = val / stops[stopIndex].cutoff
    return Color.rgb(
      minColor.red() + (maxColor.red() - minColor.red()) * positionWithinStops,
      minColor.green() + (maxColor.green() - minColor.green()) * positionWithinStops,
      minColor.blue() + (maxColor.blue() - minColor.blue()) * positionWithinStops,
    ).toString()
  }

export default colorize
