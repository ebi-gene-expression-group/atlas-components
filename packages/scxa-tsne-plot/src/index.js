import ExperimentPageView from './ExperimentPageView'

export default ExperimentPageView
