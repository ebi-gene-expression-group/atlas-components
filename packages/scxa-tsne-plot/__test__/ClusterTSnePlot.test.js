import React from 'react'
import renderer from 'react-test-renderer'
import Color from 'color'

import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import {_colourizeClusters} from '../src/ClusterTSnePlot'
import ClusterTSnePlot from '../src/ClusterTSnePlot'
import ScatterPlotLoader from '../src/plotloader/PlotLoader'

import '../src/util/MathRound'
import {randomHighchartsSeriesWithNamesAndMaxPoints} from './Utils'

Enzyme.configure({ adapter: new Adapter() })

const seriesNames = [`0`, `1`, `2`, `3`, `4`]
const maxPointsPerSeries = 1000

describe(`ClusterTSnePlot colourize function`, () => {
  test(`must not change the number of series`, () => {
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)
    expect(_colourizeClusters([], `lightgrey`)(randomSeries)).toHaveLength(seriesNames.length)
  })

  test(`must not change the number of points in each series`, () => {
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)
    _colourizeClusters([], `lightgrey`)(randomSeries).forEach((series, i) => {
      expect(series.data).toHaveLength(randomSeries[i].data.length)
    })
  })

  test(`must not dim (i.e. add a color field) any series if all are highlighted`, () => {
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)
    _colourizeClusters(seriesNames, `lightgrey`)(randomSeries).forEach((series) => {
      series.data.forEach((point) => {
        expect(point).not.toHaveProperty(`color`)
      })
    })
  })

  test(`must not dim (i.e. add a color field) any series if none are passed to highlight`, () => {
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)
    _colourizeClusters([], `lightgrey`)(randomSeries).forEach((series) => {
      series.data.forEach((point) => {
        expect(point).not.toHaveProperty(`color`)
      })
    })
  })

  test(`dims all series but the highlighted ones`, () => {
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)

    // Make sure that weâ€™re not testing the test above by highlighting at least one cluster
    let highlightRandomSeries = []
    while (highlightRandomSeries.length === 0) {
      highlightRandomSeries =
        seriesNames.reduce((acc, seriesName) =>  Math.random() > 0.5 ? acc.concat(seriesName) : acc, [])
    }

    _colourizeClusters(highlightRandomSeries, `lightgrey`)(randomSeries).forEach((series) => {
      series.data.forEach((point) => {
        if (highlightRandomSeries.includes(series.name)) {
          expect(point).not.toHaveProperty(`color`)
        } else {
          expect(point).toHaveProperty(`color`, Color(`lightgrey`).alpha(0.65).rgb().toString())
        }
      })
    })
  })
})

describe(`ClusterTSnePlot`, () => {
  test(`with no data matches snapshot`, () => {
    const onChangeK = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const tree = renderer
      .create(<ClusterTSnePlot height={500} ks={[]} selectedK={0} onChangeK={onChangeK} perplexities={[]} selectedPerplexity={0} onChangePerplexity={onChangePerplexity} loading={true} plotData={plotData}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test(`contains ScatterPlotLoader`, () => {
    const onChangeK = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const wrapper = mount(<ClusterTSnePlot height={500} ks={[]} selectedK={0} onChangeK={onChangeK} perplexities={[]} selectedPerplexity={0} onChangePerplexity={onChangePerplexity} loading={true} plotData={plotData}/>)

    expect(wrapper.find(ScatterPlotLoader).length).toBe(1)
  })
})
