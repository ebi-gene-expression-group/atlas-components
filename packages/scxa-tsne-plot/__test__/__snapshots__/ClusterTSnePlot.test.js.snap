// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterTSnePlot with no data matches snapshot 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="column medium-6"
    >
      <label>
        t-SNE Perplexity
      </label>
      <select
        onChange={[Function]}
        value={0}
      />
    </div>
    <div
      className="column medium-6"
    >
      <label>
        Number of clusters, 
        <i>
          k
        </i>
      </label>
      <select
        onChange={[Function]}
        value={0}
      />
    </div>
  </div>,
  <div
    className="row column"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className=""
    >
      <div />
    </div>
    <div />
    <div
      style={
        Object {
          "alignItems": "center",
          "background": "rgba(255,255,255,0.8)",
          "bottom": 0,
          "display": "flex",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <p>
          Loading, please wait...
        </p>
        <img
          src="test-file-stub"
        />
        <p>
          <small>
            Powered by 
            <a
              href="https://loading.io"
            >
              loading.io
            </a>
          </small>
        </p>
      </div>
    </div>
  </div>,
]
`;
